service/history/nDCTaskUtil.go

```diff
@@ -152,33 +158,47 @@ func LoadMutableStateForTask(
 }
 
 func initializeLoggerForTask(
-	shardID int32,
 	task tasks.Task,
 	logger log.Logger,
 ) log.Logger {
+	var taskEventID func(task tasks.Task) int64
+	taskCategory := task.GetCategory()
+	switch taskCategory.ID() {
+	case tasks.CategoryIDTransfer:
+		taskEventID = getTransferTaskEventID
+	case tasks.CategoryIDTimer:
+		taskEventID = getTimerTaskEventID
+	case tasks.CategoryIDVisibility:
+		// visibility tasks don't have task eventID
+		taskEventID = func(task tasks.Task) int64 { return 0 }
+	default:
+		// replication task won't reach here
+		panic(serviceerror.NewInternal("unknown task category"))
+	}
+
 	taskLogger := log.With(
 		logger,
-		tag.ShardID(shardID),
+		tag.WorkflowNamespaceID(task.GetNamespaceID()),
```

Comment:
i think it important to include shardID

Reply:
shardID tag is already attached when initializing shard context. https://github.com/temporalio/temporal/blob/master/service/history/shard/context_impl.go#L1816